### DISCUSSION TITLE
variables not being sent with the mutation

### DISCUSSION THREAD

**Discussion by smolinari** [Akryum/vue-apollo#1283] [Q&A] [answered]
Hi,

Working with v4 and composition API. 

What might I be doing wrong, if the variables aren't being sent with the mutation data? I have two refs for username and password, the refs are v-models for the input fields. The only way I can make the mutation work, is to add the variables in the template call of the mutation method. 

```html
<template>
  <q-dialog ref="dialogRef">
    <q-card square bordered class="q-pa-lg shadow-1">
      <q-card-section>
        <q-form class="q-gutter-md">
          <q-input square filled clearable v-model="username" type="email" label="username" />
          <q-input square filled clearable v-model="password" type="password" label="password" />
        </q-form>
      </q-card-section>
      <q-card-actions class="q-px-md">
        <q-btn unelevated color="light-green-7" size="lg" class="full-width" label="Login" @click="login({username: username, password: password})"/> //<--- this!!!
      </q-card-actions>
      <q-card-section class="text-center q-pa-none">
        <p class="text-grey-6">Not registered? Create an Account</p>
      </q-card-section>
    </q-card>
  </q-dialog>
</template>
``` 
I'm using Quasar Framework. Could it have something to do with using Quasar's dialog plugin?

Scott

**Comment by smolinari** [comment, marked as answer]
Ok. I've got the answer. Should have read the docs a little more carefully. (maybe the docs can be improved to not show a non-working example 🤔). At any rate, making variables a return of an arrow function fixed the problem. 

```ts
    const { mutate: login } = useMutation(gql`mutation login ($password: String!, $username: String!){
      login(password: $password, username: $username){
        accessToken
      }
    }`, 
    () => ({ // <-- this!!!!!
      variables: {
        username: username.value,
        password: password.value
      },
      update: async (cache , { data: { login: { accessToken} } }) =>{
      .....
    }))
```

Scott

**Reply by smolinari** [reply]
If anyone else runs into this issue and ends up here, let me know please with a quick comment. I'll do a PR to change the docs. 

Scott

**Reply by Akryum** [reply]
It's all documented [here](https://v4.apollo.vuejs.org/guide-composable/mutation.html#variables) already. :wink:

**Reply by smolinari** [reply]
@Akryum - Yes I know. And, I was looking at the docs more than once before I found the answer (as someone else showed me this text too). But, as normal devs would, I was looking only at the code and not the text below it explaining that the code being shown actually doesn't work. 😝 

![image](https://user-images.githubusercontent.com/3314957/142422445-28b52fc6-1477-4555-80e6-b1e57ef42f4a.png)

Scott

**Reply by Akryum** [reply]
Updated the code snippet:
![image](https://user-images.githubusercontent.com/2798204/142504412-b4f5ea53-6642-4261-b880-a0200f1783aa.png)

**Reply by smolinari** [reply]
Awesome! Thanks for the understanding @Akryum 

Scott
